### PR TITLE
Feature: Refactor CursorBasedPagination to use immutable Request functions

### PR DESCRIPTION
**Pull request by settermjd**
This PR

* [x] Refactor `CursorBasedPagination` to use immutable `Request` functions `withBeforeCursor()`, and `withAfterCursor()`

This was found on going through the code to make it more readable.

### CODE REVIEW

File: spec/Middleware/Request/Pagination/CursorBasedPaginationSpec.php [hunk 1]
```diff
@@ -57,13 +58,28 @@ public function it_will_allow_after_cursor_on_get_requests(Piston $middleware)
...
+    public function it_returns_a_payload_with_request_using_before_pagination(Piston $middleware)
     {
         $request = RequestFactory::fromGlobals()->withQueryParams(['before' => 123]);
 
         $response = $this->process($this->getPayload($request, $middleware));
         $response->shouldHaveType(Payload::class);
-        $response->getRequest()->shouldReturn($request);
+
+        $response->getRequest()->shouldHaveType(Request::class);
+        $response->getRequest()->getBeforeCursor()->shouldEqual(123);
```
Comment (localheinz) [inline spec/Middleware/Request/Pagination/CursorBasedPaginationSpec.php]: @settermjd 

How about introducing an explaining variable here instead of repeating ourself?

```php
$beforeCursor = $this->getFaker()->randomNumber();

$request = RequestFactory::fromGlobals()->withQueryParams([
    'before' => $beforeCursor,
]);

// ...

$response->getRequest()->getBeforeCursor()->shouldEqual($beforeCursor);
```

What do you think?

File: spec/Middleware/Request/Pagination/CursorBasedPaginationSpec.php [hunk 2]
```diff
@@ -57,13 +58,28 @@ public function it_will_allow_after_cursor_on_get_requests(Piston $middleware)
...
+
+    public function it_returns_a_payload_with_request_using_after_pagination(Piston $middleware)
+    {
+        $request = RequestFactory::fromGlobals()->withQueryParams(['after' => 123]);
+
+        $response = $this->process($this->getPayload($request, $middleware));
+        $response->shouldHaveType(Payload::class);
+
+        $response->getRequest()->shouldHaveType(Request::class);
+        $response->getRequest()->getAfterCursor()->shouldEqual(123);
```
Comment (localheinz) [inline spec/Middleware/Request/Pagination/CursorBasedPaginationSpec.php]: @settermjd 

How about introducing an explaining variable here instead of repeating ourself?

```php
$afterCursor = $this->getFaker()->randomNumber();

$request = RequestFactory::fromGlobals()->withQueryParams([
    'after' => $afterCursor,
]);

// ...

$response->getRequest()->getAfterCursor()->shouldEqual($afterCursor);
```

What do you think?

File: src/Middleware/Request/Pagination/CursorBasedPagination.php [hunk 3]
```diff
@@ -38,7 +38,9 @@ public function process($payload)
         $after = (isset($queryParams['after'])) ? $queryParams['after'] : null;
 
         if ($before && $after) {
-            throw new BadRequestException('You may not specify both before and after');
+            throw new BadRequestException(
+                'You may not specify both before and after'
+            );
```
Comment (localheinz) [inline src/Middleware/Request/Pagination/CursorBasedPagination.php]: @settermjd 

That line isn't really too long, is it? Can we revert that change?

File: src/Middleware/Request/Pagination/CursorBasedPagination.php [hunk 4]
```diff
@@ -46,15 +48,23 @@ public function process($payload)
...
             if ($before) {
-                $request->setBeforeCursor($before);
+                $request = $request->withBeforeCursor($before);
 
-                return $payload;
+                return new Payload(
+                    $payload->getSubject(),
+                    $request,
+                    $payload->getResponse()
+                );
```
Comment (localheinz) [inline src/Middleware/Request/Pagination/CursorBasedPagination.php]: @settermjd 

How about inlining `$request` here?

```php
if ($before) {
    return new Payload(
        $payload->getSubject(),
        $request->withBeforeCursor($before),
        $payload->getResponse()
    );
}
```

File: src/Middleware/Request/Pagination/CursorBasedPagination.php [hunk 5]
```diff
@@ -46,15 +48,23 @@ public function process($payload)
...
             if ($after) {
-                $request->setAfterCursor($after);
+                $request = $request->withAfterCursor($after);
 
-                return $payload;
+                return new Payload(
+                    $payload->getSubject(),
+                    $request,
+                    $payload->getResponse()
+                );
```
Comment (localheinz) [inline src/Middleware/Request/Pagination/CursorBasedPagination.php]: @settermjd 

How about inlining `$request` here?

```php
if ($after) {
    return new Payload(
        $payload->getSubject(),
        $request->withAfterCursor($after),
        $payload->getResponse()
    );
}
```

File: src/Router/MiddlewareStrategy.php [hunk 6]
```diff
@@ -27,19 +27,37 @@ class MiddlewareStrategy extends RequestResponseStrategy implements StrategyInte
      *
      * @return ApiResponse
      */
-    public function dispatch(callable $controller, array $vars = [], Route $route = null)
-    {
-        if ($group = $route->getParentGroup()) {
+    public function dispatch(
+        callable $controller,
+        array $vars = [],
+        Route $route = null
```
Comment (localheinz) [inline src/Router/MiddlewareStrategy.php]: @settermjd 

This line isn't really too long, is it? Can we revert the change?

File: src/Router/MiddlewareStrategy.php [hunk 7]
```diff
@@ -27,19 +27,37 @@ class MiddlewareStrategy extends RequestResponseStrategy implements StrategyInte
...
+    ) {
+        $group = $route->getParentGroup();
+
+        if ($group) {
             $this->response = (new PipelineProcessor())
-                ->handlePayload(new Payload($group, $this->request, $this->response))
+                ->handlePayload(
+                    new Payload(
```
Comment (localheinz) [inline src/Router/MiddlewareStrategy.php]: @settermjd 

Can we revert all unrelated changes, please?

Pull requests are a lot easier to review if they don't contain unrelated changes - these changes here are only changes related to coding style. If you like to change the style, please do so in a separate PR (that only contains coding style changes, then).

It's basically one application of the Single Responsibility Principle! :smiley:

Comment (settermjd) [inline src/Router/MiddlewareStrategy.php]: :+1: fair call

**Comment by settermjd** [comment]
@localheinz all feedback addressed.
